### ISSUE TITLE
Error when building Ubuntu Xenial template

### ISSUE THREAD

**Issue by austinjmorlan** [QubesOS/qubes-issues#3871]
### Qubes OS version:

4.0

### Affected component(s): 

core-agent-linux

---

### Steps to reproduce the behavior:

On Fedora 26:
```shell
sudo dnf install git createrepo rpm-build make wget rpmdevtools dialog rpm-sign gnupg dpkg-dev debootstrap python2-sh perl-Digest-MD5 perl-Digest-SHA
wget https://keys.qubes-os.org/keys/qubes-master-signing-key.asc
gpg --import qubes-master-signing-key.asc 
gpg --edit-key 36879494
fpr
trust
5
y
q
wget https://keys.qubes-os.org/keys/qubes-developers-keys.asc
gpg --import qubes-developers-keys.asc
git clone git://github.com/QubesOS/qubes-builder.git
cd qubes-builder
git tag -v `git describe`
mkdir -p keyrings/git
cp ~/.gnupg/pubring.gpg ~/.gnupg/trustdb.gpg keyrings/git
./setup
Yes
Yes
4.0
Stable
Yes
Select builder-fedora and builder-debian
Yes
Select only xenial
make install-deps
make get-sources
make qubes-vm
```

### Expected behavior:

Ubuntu template should build without error.

### Actual behavior:

Build fails with error.

End of [core-agent-linux-vm-xenial.log](https://github.com/QubesOS/qubes-issues/files/1962154/core-agent-linux-vm-xenial.log):

```shell
# Install system dropins
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/lib/systemd/system
install -m 0644 vm-systemd/qubes-*.socket /home/user/qubes-src/core-agent-linux/debian/tmp/lib/systemd/system/
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install network/setup-ip /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install network/qubes-fix-nm-conf.sh /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install network/network-manager-prepare-conf-dir /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/NetworkManager/dispatcher.d/
install network/{qubes-nmhook,30-qubes-external-ip} /home/user/qubes-src/core-agent-linux/debian/tmp/etc/NetworkManager/dispatcher.d/
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/NetworkManager/conf.d
install -m 0644 network/nm-30-qubes.conf /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/NetworkManager/conf.d/30-qubes.conf
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/xdg/autostart
install -m 0755 network/show-hide-nm-applet.sh /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/
install -m 0644 network/show-hide-nm-applet.desktop /home/user/qubes-src/core-agent-linux/debian/tmp/etc/xdg/autostart/00-qubes-show-hide-nm-applet.desktop
install -D -m 0644 vm-systemd/qubes-firewall.service vm-systemd/qubes-iptables.service vm-systemd/qubes-updates-proxy.service /home/user/qubes-src/core-agent-linux/debian/tmp/lib/systemd/system/
install -D -m 0755 network/qubes-iptables /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/init/qubes-iptables
install -D -m 0644 vm-systemd/qubes-core-agent-linux.tmpfiles \
	/home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/tmpfiles.d/qubes-core-agent-linux.conf
mkdir -p /home/user/qubes-src/core-agent-linux/debian/tmp/usr/sbin
mv /home/user/qubes-src/core-agent-linux/debian/tmp/usr/bin/qubes-firewall /home/user/qubes-src/core-agent-linux/debian/tmp/usr/sbin/qubes-firewall
install -D network/qubes-setup-dnat-to-ns /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/qubes-setup-dnat-to-ns
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/dhclient.d
ln -s /usr/lib/qubes/qubes-setup-dnat-to-ns /home/user/qubes-src/core-agent-linux/debian/tmp/etc/dhclient.d/qubes-setup-dnat-to-ns.sh
install -D network/vif-route-qubes /home/user/qubes-src/core-agent-linux/debian/tmp/etc/xen/scripts/vif-route-qubes
install -D network/vif-qubes-nat.sh /home/user/qubes-src/core-agent-linux/debian/tmp/etc/xen/scripts/vif-qubes-nat.sh
install -m 0644 -D network/tinyproxy-updates.conf /home/user/qubes-src/core-agent-linux/debian/tmp/etc/tinyproxy/tinyproxy-updates.conf
install -m 0644 -D network/updates-blacklist /home/user/qubes-src/core-agent-linux/debian/tmp/etc/tinyproxy/updates-blacklist
install -m 0755 -D network/iptables-updates-proxy /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/iptables-updates-proxy
install -m 0400 -D network/iptables /home/user/qubes-src/core-agent-linux/debian/tmp/etc/qubes/iptables.rules
install -m 0400 -D network/ip6tables /home/user/qubes-src/core-agent-linux/debian/tmp/etc/qubes/ip6tables.rules
install -m 0400 -D network/ip6tables-enabled /home/user/qubes-src/core-agent-linux/debian/tmp/etc/qubes/ip6tables-enabled.rules
install -m 0755 -D qubes-rpc/qubes.UpdatesProxy /home/user/qubes-src/core-agent-linux/debian/tmp/etc/qubes-rpc/qubes.UpdatesProxy
mkdir -p /home/user/qubes-src/core-agent-linux/debian/tmp/etc/apt/sources.list.d
sed -e "s/@DIST@/`lsb_release -cs`/" misc/qubes-r4.list.in > /home/user/qubes-src/core-agent-linux/debian/tmp/etc/apt/sources.list.d/qubes-r4.list
install -D -m 644 misc/qubes-archive-keyring.gpg /home/user/qubes-src/core-agent-linux/debian/tmp/etc/apt/trusted.gpg.d/qubes-archive-keyring.gpg
install -D -m 644 network/00notify-hook /home/user/qubes-src/core-agent-linux/debian/tmp/etc/apt/apt.conf.d/00notify-hook
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/sysctl.d
install -m 644 network/80-qubes.conf /home/user/qubes-src/core-agent-linux/debian/tmp/etc/sysctl.d/
install -D -m 644 misc/profile.d_qt_x11_no_mitshm.sh /home/user/qubes-src/core-agent-linux/debian/tmp/etc/profile.d/qt_x11_no_mitshm.sh
install -D -m 440 misc/sudoers.d_umask /home/user/qubes-src/core-agent-linux/debian/tmp/etc/sudoers.d/umask
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/pam.d
install -m 0644 misc/pam.d_su.qubes /home/user/qubes-src/core-agent-linux/debian/tmp/etc/pam.d/su.qubes
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/needrestart/conf.d
install -D -m 0644 misc/50_qubes.conf /home/user/qubes-src/core-agent-linux/debian/tmp/etc/needrestart/conf.d/50_qubes.conf
install -D -m 0644 misc/grub.qubes /home/user/qubes-src/core-agent-linux/debian/tmp/etc/default/grub.d/30-qubes.cfg
install -D -m 0644 misc/apt-conf-70no-unattended /home/user/qubes-src/core-agent-linux/debian/tmp/etc/apt/apt.conf.d/70no-unattended
mkdir -p /home/user/qubes-src/core-agent-linux/debian/tmp/etc/systemd/system/
install -m 0644 vm-systemd/haveged.service  /home/user/qubes-src/core-agent-linux/debian/tmp/etc/systemd/system/
make[2]: Leaving directory '/home/user/qubes-src/core-agent-linux'
make -C qrexec install
make[2]: Entering directory '/home/user/qubes-src/core-agent-linux/qrexec'
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/etc/qubes-rpc
install -d /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes /home/user/qubes-src/core-agent-linux/debian/tmp/usr/bin
install qrexec-agent /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes
install qrexec-client-vm /home/user/qubes-src/core-agent-linux/debian/tmp/usr/bin
ln -s ../../bin/qrexec-client-vm /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/qrexec-client-vm
ln -s ../../bin/qrexec-client-vm /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes/qrexec_client_vm
install qrexec-fork-server /home/user/qubes-src/core-agent-linux/debian/tmp/usr/bin
install qubes-rpc-multiplexer /home/user/qubes-src/core-agent-linux/debian/tmp/usr/lib/qubes
install -D -m 0644 qrexec.pam /home/user/qubes-src/core-agent-linux/debian/tmp/etc/pam.d/qrexec
make[2]: Leaving directory '/home/user/qubes-src/core-agent-linux/qrexec'
make[1]: Leaving directory '/home/user/qubes-src/core-agent-linux'
   debian/rules override_dh_install
make[1]: Entering directory '/home/user/qubes-src/core-agent-linux'
dh_install --fail-missing
dh_install: lib/systemd/system/exim4.service.d/30_qubes.conf exists in debian/tmp but is not installed to anywhere
dh_install: lib/systemd/system/netfilter-persistent.service.d/30_qubes.conf exists in debian/tmp but is not installed to anywhere
dh_install: lib/systemd/system/avahi-daemon.service.d/30_qubes.conf exists in debian/tmp but is not installed to anywhere
dh_install: missing files, aborting
debian/rules:28: recipe for target 'override_dh_install' failed
make[1]: *** [override_dh_install] Error 2
make[1]: Leaving directory '/home/user/qubes-src/core-agent-linux'
debian/rules:12: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary gave error exit status 2
make[2]: *** [/home/austin/qubes-builder/qubes-src/builder-debian/Makefile.qubuntu:215: dist-package] Error 
```

### General notes:

I'm able to fix this by reverting the changes made to `Makefile.builder` in https://github.com/QubesOS/qubes-core-agent-linux/pull/71/commits/80b5c942069710f75aa7faea141e31c1cba60b33 and https://github.com/QubesOS/qubes-core-agent-linux/pull/70/commits/54867b6eabed3a6dca0e932ff77f67aedfe43e03, but I'm not sure if that's a correct solution, or why those changes were made to that file in the first place.

---

**Comment by andrewdavidwong** [comment]
Duplicate of #3308

**Comment by andrewdavidwong** [comment]
This appears to be a duplicate of an existing issue. If you believe this is not really a duplicate, please leave a comment briefly explaining why. We'll be happy to take another look and, if appropriate, reopen this issue. Thank you.

**Comment by austinjmorlan** [comment]
I don't believe this to be a duplicate of #3308 .

#3308 was closed with https://github.com/QubesOS/qubes-core-agent-linux/pull/71, but it seems that the changes in that PR are what's causing my current build to fail.

**Comment by ptitdoc** [comment]
Same problem here. The logic used to avoid installing some of the systemd dropins in https://github.com/QubesOS/qubes-core-agent-linux/pull/71/commits/80b5c942069710f75aa7faea141e31c1cba60b33 is apparently broken. As a temporary workaround, one can remove installing the broken files from core-agent-linux main Makefile directly:
```
 diff --git a/Makefile b/Makefile
index a99ef4d..1b721fd 100644
--- a/Makefile
+++ b/Makefile
@@ -84,9 +84,9 @@ ifeq ($(shell lsb_release -is), Ubuntu)
     SYSTEM_DROPINS := $(strip $(patsubst crond.service, cron.service, $(SYSTEM_DROPINS)))
     SYSTEM_DROPINS += anacron.service
     SYSTEM_DROPINS += anacron-resume.service
-    SYSTEM_DROPINS += netfilter-persistent.service
-    SYSTEM_DROPINS += exim4.service
-    SYSTEM_DROPINS += avahi-daemon.service
+    #SYSTEM_DROPINS += netfilter-persistent.service
+    #SYSTEM_DROPINS += exim4.service
+    #SYSTEM_DROPINS += avahi-daemon.service
 
 endif
```

**Comment by gasull** [comment]
I can confirm @ptitdoc 's workaround worked.  You need to edit the file `./qubes-src/core-agent-linux/Makefile`.  Unfortunately this led to new errors and I continued commenting out more code in `./Makefile`:
```diff
diff --git a/Makefile b/Makefile
index 8d2af74..bb77dea 100644
--- a/Makefile
+++ b/Makefile
@@ -442,22 +442,22 @@ mostlyclean:: clean-chroot clean-rpms clean
        fi
 
 .PHONY: iso iso.clean-repos iso.copy-rpms iso.copy-template-rpms
-iso.clean-repos:
-       @echo "-> Preparing for ISO build..."
-       @$(MAKE) -s -C $(SRC_DIR)/$(INSTALLER_COMPONENT) clean-repos
+#iso.clean-repos:
+#      @echo "-> Preparing for ISO build..."
+#      @$(MAKE) -s -C $(SRC_DIR)/$(INSTALLER_COMPONENT) clean-repos
 
 iso.copy-rpms: $(COMPONENTS_NO_TPL_BUILDER:%=iso.copy-rpms.%)
 
 iso.copy-rpms.%: COMPONENT=$*
 iso.copy-rpms.%: REPO=$(SRC_DIR)/$(COMPONENT)
-iso.copy-rpms.%: $(SRC_DIR)/%/Makefile.builder
-       @echo "--> Copying $(COMPONENT) RPMs..."
-       @$(MAKE) --no-print-directory -f Makefile.generic \
-               PACKAGE_SET=dom0 \
-               DIST=$(DIST_DOM0) \
-               COMPONENT=$(COMPONENT) \
-               UPDATE_REPO=$(BUILDER_DIR)/$(SRC_DIR)/$(INSTALLER_COMPONENT)/yum/qubes-dom0 \
-               update-repo
+#iso.copy-rpms.%: $(SRC_DIR)/%/Makefile.builder
+#      @echo "--> Copying $(COMPONENT) RPMs..."
+#      @$(MAKE) --no-print-directory -f Makefile.generic \
+#              PACKAGE_SET=dom0 \
+#              DIST=$(DIST_DOM0) \
+#              COMPONENT=$(COMPONENT) \
+#              UPDATE_REPO=$(BUILDER_DIR)/$(SRC_DIR)/$(INSTALLER_COMPONENT)/yum/qubes-dom0 \
+#              update-repo
 
 iso.copy-rpms.%: $(SRC_DIR)/%/Makefile

            @echo "--> Copying $(COMPONENT) RPMs..."
```

Then this led to a final error where I am stuck:

```
make: *** No rule to make target 'iso.copy-rpms.builder-rpm', needed by 'iso.copy-rpms'.  Stop.
```

**Comment by marmarek** [comment]
@gasull if you're building just template, you don't need to execute `make iso`. If you got the point where it tries to build one, the template is already complete, check `qubes-src/linux-template-builder/rpm/noarch` directory.

**Comment by gasull** [comment]
I tried again following the [instructions for Archlinux](https://www.qubes-os.org/doc/building-archlinux-template/), but also installing `perl-Digest-MD5` and `perl-Digest-SHA` as [described by the OP](https://github.com/QubesOS/qubes-issues/issues/3871#issue-319036355) and executing `make switch-branch` as [described](https://groups.google.com/d/msg/qubes-users/rRir2d86hOM/XpwgH04uAgAJ) by @unman.

That led to the error described by the OP, and then I used the [workaround](https://github.com/QubesOS/qubes-issues/issues/3871#issuecomment-390459578) described by @ptitdoc.

The end result is that the TemplateVM is installed but it does't show the applications in them.  Trying to refresh the applications in the settings never completes.  I cannot even run `xterm` with "Run command in qube".  The qube does start and shutdown, though.

I also built Trusty and it has the same problem plus it doesn't even shutdown.

**Comment by Bufil** [comment]
Same as @gasull here too.
Maybe it has something to do with #4067 ?
In Whonix14 i can refresh the applications but nothing starts.

**Comment by Bufil** [comment]
#4067 had to do with #3666.
So that problem should be different.

**Comment by Bufil** [comment]
`sudo xl console xenial-desktop` works
Output of `systemctl status qubes-gui-agent`:
```
root@xenial-desktop:~# systemctl status qubes-gui-agent
● qubes-gui-agent.service - Qubes GUI Agent
   Loaded: loaded (/lib/systemd/system/qubes-gui-agent.service; enabled; vendor 
   Active: active (running) since Sat 2018-07-14 14:42:44 CEST; 3min 24s ago
  Process: 661 ExecStartPre=/bin/touch /var/run/console/user (code=exited, statu
  Process: 655 ExecStartPre=/bin/mkdir -p /var/run/console (code=exited, status=
 Main PID: 664 (qubes-gui)
   CGroup: /system.slice/qubes-gui-agent.service
           └─664 /usr/bin/qubes-gui

Jul 14 14:42:44 xenial-desktop systemd[1]: Starting Qubes GUI Agent...
Jul 14 14:42:44 xenial-desktop systemd[1]: Started Qubes GUI Agent.
Jul 14 14:42:44 xenial-desktop qubes-gui[664]: Waiting on /var/run/xf86-qubes-so
Jul 14 14:42:44 xenial-desktop su[702]: Successful su for user by root
Jul 14 14:42:44 xenial-desktop su[702]: + ??? root:user
Jul 14 14:42:44 xenial-desktop su[702]: pam_unix(su:session): session opened for
Jul 14 14:42:45 xenial-desktop qubes-gui[664]: Ok, somebody connected.
Jul 14 14:42:46 xenial-desktop qubes-gui[664]: got num_mfn=0x0 for window 0xc000
Jul 14 14:42:46 xenial-desktop qubes-gui[664]: error reading WM_NORMAL_HINTS
Jul 14 14:42:46 xenial-desktop qubes-gui[664]: error reading WM_NORMAL_HINTS
```
